zjit/src/hir.rs

```diff
@@ -912,9 +923,27 @@ impl Function {
     fn infer_types(&mut self) {
         // Reset all types
         self.insn_types.fill(types::Empty);
-        for param in &self.blocks[self.entry_block.0].params {
+
+        // Index of the rest parameter for comparison below
+        let rest_param_idx = *self.rest_param_idx.get_or_init(|| {
+            if !self.iseq.is_null() && unsafe { get_iseq_flags_has_rest(self.iseq) } {
+                let opt_num = unsafe { get_iseq_body_param_opt_num(self.iseq) };
+                let lead_num = unsafe { get_iseq_body_param_lead_num(self.iseq) };
+                opt_num + lead_num
+            } else {
+                -1
+            }
+        });
+
+        // Fill parameter types
+        for (idx, param) in self.blocks[self.entry_block.0].params.iter().enumerate() {
             // We know that function parameters are BasicObject or some subclass
             self.insn_types[param.0] = types::BasicObject;
+
+            // Rest parameters are always ArrayExact
+            if let Ok(true) = c_int::try_from(idx).map(|idx| idx == rest_param_idx) {
+                self.insn_types[param.0] = types::ArrayExact;
```

Comment:
Can you store a `param_types` vec on `Function` instead of `rest_param_idx`? If you'd rather not do it in this PR I can stack a new commit.

Reply:
(& initialize in HIR builder, not OnceCell)